ledger-service/db-backend/src/test/scala/com/digitalasset/http/dbbackend/QueriesSpec.scala

```diff
@@ -3,33 +3,75 @@
 
 package com.daml.http.dbbackend
 
+import doobie.Fragment
+import doobie.implicits._
 import org.scalatest.matchers.should.Matchers
+import org.scalatest.prop.TableDrivenPropertyChecks
 import org.scalatest.wordspec.AnyWordSpec
-import org.scalatestplus.scalacheck.ScalaCheckDrivenPropertyChecks
 import scala.collection.immutable.Seq
+import scalaz.\/
 
-class QueriesSpec extends AnyWordSpec with Matchers with ScalaCheckDrivenPropertyChecks {
-  "uniqueSets" should {
-    import Queries.uniqueSets
-    "return only a map if given and non-empty" in forAll { (m: Map[Int, Int], c: (Int, Int)) =>
-      val nem = m + c
-      uniqueSets(nem) should ===(Seq(nem))
-    }
+class QueriesSpec extends AnyWordSpec with Matchers with TableDrivenPropertyChecks {
+  import QueriesSpec._
 
-    "return empty for empty" in {
-      uniqueSets(Seq.empty[(Int, Int)]) should ===(Seq.empty)
-    }
+  "projectedIndex" should {
+    import Queries.{projectedIndex, SurrogateTpId}
 
-    "return n elements in order if all given the same key" in forAll { xs: Seq[Int] =>
-      uniqueSets(xs map ((42, _))) should ===(xs map (x => Map(42 -> x)))
-    }
+    final case class Reordered(s_2_13: Fragment, s_2_12: Fragment)
+    def Agnostic(fragment: Fragment) = Reordered(fragment, fragment)
+
+    val cases = Table(
+      ("map", "projection"),
+      (
+        Seq((SurrogateTpId(55L), fr"foo") -> 0),
+        Agnostic(sql"CASE WHEN ( tpid = ${55L}) THEN ${0}||''  ELSE NULL END "),
+      ),
+      (
+        Seq((SurrogateTpId(55L), fr"foo") -> 0, (SurrogateTpId(66L), fr"foo") -> 1),
+        Reordered(
+          s_2_12 =
+            sql"CASE WHEN ( tpid = ${55L}) THEN ${0}||''  WHEN ( tpid = ${66L}) THEN ${1}||''  ELSE NULL END ",
+          s_2_13 =
+            sql"CASE WHEN ( tpid = ${66L}) THEN ${1}||''  WHEN ( tpid = ${55L}) THEN ${0}||''  ELSE NULL END ",
+        ),
+      ),
+      (
+        Seq(
+          (SurrogateTpId(55L), fr"foo") -> 0,
+          (SurrogateTpId(55L), fr"bar") -> 1,
+          (SurrogateTpId(66L), fr"baz") -> 2,
+        ),
+        Reordered(
```

Comment:
Should we just use an ordered map instead? This really shouldn’t be our performance bottleneck and that seems like a much nicer solution than hardcoding the hash order.

Reply:
Yeah, we can, I'm just being stubborn about my instances.

Reply:
fc7db44d36b

Reply:
nice, thank you